### PR TITLE
Feature/ball handle pos control

### PR DESCRIPTION
**Pull request by ThijsLuttikhuis**
- [x] The tests are passing

### Summary
Give a short summary of what the changes are and why you made them. If applicable, apply screenshots as well. 

### Related issues
If there are related issues, reference them here.

fixes ...  
fixes ...

### Comments for the reviewer
If there are comments for the reviewer, leave them here

### Checklist for the reviewer
This acts as a guide to help you and/or the reviewer to analyze the code. 

**functionality**  
- Is the code is properly tested?
- Is the code working as expected
- Is the code is understandable and easy to read
- Are proper abstractions made? (no duplicate code, nice use of functions/objects)

**style**  
- Variables/objects/classes have proper names
- Proper use of variables: nu unused vars, no magic numbers
- No possible memory leaks. Mutexes properly used when multithreading? possible pointer issues?
- Are smart pointers used?
- unused variables?
- magic numbers?
- well documented?
- No unused imports?
- No leftover code / unused comments
- Proper indentation
- Proper use of namespaces

### CODE REVIEW

File: roboteam_ai/src/skills/gotopos/GTPWithBall.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#include <utility>
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/gotopos/GTPWithBall.cpp]: unnecessary includes here

File: roboteam_ai/src/skills/gotopos/GTPWithBall.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+Skill::Status GTPWithBall::onUpdate() {
+    updateTarget();
+    command = ballHandlePosControl.getPosVelAngle(robot, targetPos, targetAngle).makeRobotCommand();
+    publishRobotCommand();
+    return Status::Running;
+}
+
+void GTPWithBall::onTerminate(Skill::Status s) {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/skills/gotopos/GTPWithBall.cpp]: Typically on terminate we don't want to send robot commands

Comment (rolfvdhulst) [reply roboteam_ai/src/skills/gotopos/GTPWithBall.cpp]: Unless it is for ballplacement or so etc.

File: roboteam_ai/src/world/Robot.cpp [hunk 3]
```diff
@@ -90,6 +90,20 @@ void Robot::setGenevaState(unsigned char state) {
     else std::cout << "setting invalid geneva state (" << state << ") for robot with id " << id << std::endl;
 }
 
+const Robot Robot::clone() const {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/Robot.cpp]: Robot newBot=oldBot::clone() is identical to Robot newBot=oldBot so this is a redundant function.

File: roboteam_ai/src/world/World.cpp [hunk 4]
```diff
@@ -335,7 +335,7 @@ const World::RobotPtr World::getFutureRobot(const RobotPtr &robotPtr, double tim
 }
 
 const Robot World::getFutureRobot(const Robot &robot, double time) {
-    Robot futureRobot = robot;
+    Robot futureRobot = robot.clone();
```
Comment (rolfvdhulst) [inline roboteam_ai/src/world/World.cpp]: Robot futureBot=robot should just work?

File: roboteam_ai/src/control/positionControllers/RobotCommand.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        unsigned char dribbler = 0;
+        unsigned char geneva = 3;
+        bool kicker = false;
+        double kickerVel = 0;
+        bool kickerForced = false;
+        bool chipper = false;
+        double chipperVel = 0;
+        bool chipperForced = false;
+
+        roboteam_msgs::RobotCommand makeRobotCommand() const {
```
[the remark this answers is not in the document]
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/positionControllers/RobotCommand.h]: agreed

File: roboteam_ai/src/control/positionControllers/RobotCommand.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+#define ROBOTEAM_AI_ROBOTCOMMAND_H
+
+#include "roboteam_utils/Vector2.h"
+#include "roboteam_utils/Angle.h"
+
+#include "roboteam_msgs/RobotCommand.h"
+
+namespace rtt {
+
+class RobotCommand {
```
[the remark this answers is not in the document]
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/control/positionControllers/RobotCommand.h]: yes it is kind of a duplicate, but it is not autogenerated so its functionality can be improved

File: roboteam_ai/src/world/World.cpp (hunk 4 above)
Comment (ThijsLuttikhuis) [reply roboteam_ai/src/world/World.cpp]: not sure, it probably does

File: roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+RobotCommand BallHandlePosControl::getPosVelAngle(const RobotPtr &r,
+        const Vector2 &targetP, const Angle &targetA) {
+
+    // update variables
+    updateVariables(r, targetP, targetA);
+
+    // check for ball
+    if (! ball) {
+        if (Constants::SHOW_BALL_HANDLE_DEBUG_INFO()) {
+            std::cout << "Can't control the ball with no ball, you stupid" << std::endl;
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: remove the last two words, we don't want to promote profanity

Comment (rolfvdhulst) [reply roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: As in we are actually nice people

File: roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+            forwardsProgress = F_start;
+            return {};
+        }
+        }
+    }
+
+    case defaultTravel:break;
+    }
+
+    std::cout << "travel case not handled" << std::endl;
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: This print shouldn't be necessary as compiler should explicitly warn you in this case. Can do no big harm though.

File: roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+    return command;
+}
+
+RobotCommand BallHandlePosControl::B_sendDribbleBackwardsCommand() {
+    RobotCommand command;
+    command.dribbler = 1;
+    command.angle = B_lockedAngle;
+    command.vel = B_lockedAngle.toVector2(- maxBackwardsVelocity);
+
+    if (control::ControlUtils::distanceToLine(robot->pos, B_backwardsDribbleLine.first, B_backwardsDribbleLine.second)
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: This needs some comments, hard to follow

Comment (rolfvdhulst) [reply roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: or atleast one

File: roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp [hunk 10]
```diff
@@ -0,0 +1,514 @@
...
+        }
+        return;
+    }
+    case B_dribbling: {
+        if (! robot->hasBall()) {
+            backwardsProgress = B_approaching;
+            return;
+        }
+        B_count ++;
+        if (B_count > 25) {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: This is quite an important constant, put it in the header IMO.

File: roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp [hunk 11]
```diff
@@ -0,0 +1,514 @@
...
+        }
+        return;
+    }
+    case B_dribbleBackwards: {
+        if (! robot->hasBall()) {
+            backwardsProgress = B_approaching;
+            return;
+        }
+        Angle offsetAngle = (robot->pos - finalTargetPos).toAngle() - robot->angle;
+        if (offsetAngle > M_PI*0.05) {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: Put this in the header too

File: roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp [hunk 12]
```diff
@@ -0,0 +1,514 @@
...
+        if ((robot->pos - finalTargetPos).length2() < errorMargin*errorMargin) {
+            return;
+        }
+        backwardsProgress = B_start;
+    }
+    }
+}
+
+void BallHandlePosControl::updateForwardsProgress() {
+    if (Constants::SHOW_BALL_HANDLE_DEBUG_INFO()) {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: Make this print one function

Comment (rolfvdhulst) [reply roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: it distracts fromt he point here

File: roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp [hunk 13]
```diff
@@ -0,0 +1,514 @@
...
+    // acceleration limiter
+    double maxAcc = control::ControlUtils::calculateMaxAcceleration(command.vel, command.angle);
+    command.vel = control::ControlUtils::accelerationLimiter(command.vel, maxAcc, previousVelocity.length());
+    previousVelocity = command.vel;
+    return command;
+}
+
+RobotCommand BallHandlePosControl::goToBall(bool robotDoesNotHaveBall, bool robotIsTooFarFromBall) {
+    if (Constants::SHOW_BALL_HANDLE_DEBUG_INFO()) {
+        std::cout << "we do not have a ball yet idiot" << std::endl;
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: No profanity please.

File: roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+
+RobotCommand BallHandlePosControl::F_sendApproachCommand() {
+    RobotCommand command;
+    command.dribbler = 0;
+    command.vel = ballToRobot.stretchToLength(maxForwardsVelocity);
+    command.angle = F_lockedAngle;
+    return command;
+}
+
+RobotCommand BallHandlePosControl::F_sendDribbleForwardsCommand() {
```
Comment (rolfvdhulst) [inline roboteam_ai/src/control/positionControllers/BallHandlePosControl.cpp]: This needs a small comment to explain the maths a bit.

File: roboteam_ai/src/skills/gotopos/GTPWithBall.cpp (hunk 2 above)
Comment (rolfvdhulst) [reply roboteam_ai/src/skills/gotopos/GTPWithBall.cpp]: Even then, preferably not.

File: roboteam_ai/src/Switches.cpp [hunk 15]
```diff
@@ -256,7 +257,7 @@ bt::Node::Ptr Switches::leafSwitch(std::string name, bt::Blackboard::Ptr propert
     map["PenaltyKeeper"] = std::make_shared<rtt::ai::PenaltyKeeper>(name, properties);
     map["FreeKickFormation"] = std::make_shared<rtt::ai::FreeKickFormation>(name, properties);
     map["DefendFreeKick"] = std::make_shared<rtt::ai::DefendFreeKick>(name, properties);
-
+    map["GTPWithBall"] = std::make_shared<rtt::ai::GTPWithBall>(name, properties);
```
[the remark this answers is not in the document]
Comment (Barisimre) [reply roboteam_ai/src/Switches.cpp]: i like you
